src/librustc_mir/borrow_check/nll/region_infer/error_reporting/outlives_suggestion.rs

```diff
@@ -0,0 +1,256 @@
+    /// suggestable.
+    crate fn intermediate_suggestion(
+        &mut self,
+        errctx: &ErrorReportingCtx<'_, '_, '_>,
+        errci: &ErrorConstraintInfo,
+        renctx: &mut RegionErrorNamingCtx,
+        diag: &mut DiagnosticBuilder<'_>,
+    ) {
+        // Emit an intermediate note.
+        let fr_name = self.region_vid_to_name(errctx, renctx, errci.fr);
+        let outlived_fr_name = self.region_vid_to_name(errctx, renctx, errci.outlived_fr);
+
+        if let (Some(fr_name), Some(outlived_fr_name)) = (fr_name, outlived_fr_name) {
+            if let RegionNameSource::Static = outlived_fr_name.source {
+                diag.help(&format!(
+                    "consider replacing `{}` with `'static`", fr_name
+                ));
+            } else {
+                diag.help(&format!(
+                    "consider adding the following bound: `{}: {}`",
+                    fr_name, outlived_fr_name
+                ));
+            }
+        }
+    }
+
+    /// If there is a suggestion to emit, add a diagnostic to the buffer. This is the final
+    /// suggestion including all collected constraints.
+    crate fn add_suggestion<'tcx>(
+        &self,
+        body: &Body<'tcx>,
+        rinfcx: &RegionInferenceContext<'tcx>,
+        infcx: &InferCtxt<'_, 'tcx>,
+        errors_buffer: &mut Vec<Diagnostic>,
+        renctx: &mut RegionErrorNamingCtx,
+    ) {
+        // No constraints to add? Done.
+        if self.constraints_to_add.is_empty() {
+            debug!("No constraints to suggest.");
+            return;
+        }
+
+        // If there is only one constraint to suggest, then we already suggested it in the
+        // intermediate suggestion above.
+        if self.constraints_to_add.len() == 1 {
+            debug!("Only 1 suggestion. Skipping.");
+            return;
+        }
+
+        // Create a new diagnostic.
+        let mut diag = DiagnosticBuilder::new(
+            infcx.tcx.sess.diagnostic(),
+            Level::Help,
+            "the following changes may resolve your lifetime errors",
+        );
+
+        let mir_span = infcx.tcx.def_span(self.mir_def_id);
+
+        // Make sure that there is actually a suggestion to emit.
+        let mut ndiags = 0;
+
+        // We want this message to appear after other messages on the mir def.
+        diag.sort_span = mir_span.shrink_to_hi();
+
+        // Keep track of variables that we have already suggested unifying so that we don't print
+        // out silly duplicate messages.
+        let mut unified_already = FxHashSet::default();
+
+        let errctx = ErrorReportingCtx {
+            rinfcx, infcx, body,
+            mir_def_id: self.mir_def_id,
+
+            // We should not be suggesting naming upvars, so we pass in a dummy set of upvars that
+            // should never be used.
+            upvars: &[]
+        };
+
+        for (fr, outlived) in &self.constraints_to_add {
```

Comment:
This loop has a non deterministic iteration order as `constraints_to_add` is an `FxHashMap`. Because it contains a `diag.help`, that diagnostic will be non deterministic too.

Reply:
Ah, excellent! I thought i missed one somewhere

Reply:
Somebody should make a `FxHashMap` wrapper which doesnt provide iteration support, but only insert/remove/get/get_mut and convertion to a sorted vec

Reply:
@bjorn3 Perhaps create an issue for that?

Reply:
Filled https://github.com/rust-lang/rust/issues/63713